File: 52. N-Queens II.py
class Solution(object):
    def totalNQueens(self, n):
        """
        :type n: int
        :rtype: int
        """
        self.res = 0
        self.DFS([], n)
        return self.res
        
    def check(self, path, i):
        for j in xrange(len(path)):
            if i == path[j] or len(path)-j == abs(path[j] - i):
                return False
        return True
    def DFS(self, path, n):
        if len(path) == n:
            self.res += 1
        for i in xrange(n):
            if self.check(path, i):
                self.DFS(path + [i], n)

File: 52. N-Queens II.cpp
class Solution {
private:
    int res = 0;
public:
    int totalNQueens(int n) {
        DFS({}, n);
        return res;
    }
    bool check(vector <int> path, int i)
    {
        for (int j = 0; j < path.size(); j++)
        {
            if (path[j] == i || path.size() - j == abs(path[j] - i))
                return false;
        }
        return true;
    }
    void DFS(vector <int> path, int& n)
    {
        if (path.size() == n)
        {
            res ++;
            return;
        }
        for (int i = 0; i < n; i++)
        {
            if (check(path, i))
            {
                path.push_back(i);
                DFS(path, n);
                path.pop_back();
            }
        }
    }
};
